x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go

```diff
@@ -55,170 +48,138 @@ func NewPipelineManager(logCfg *common.Config) *PipelineManager {
+// checkIfHashExists is a short atomic function to see if a pipeline alread exists inside the PM. Thread-safe.
+func (pm *PipelineManager) checkIfHashExists(logOptsConfig map[string]string) bool {
+	hashstring := makeConfigHash(logOptsConfig)
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	_, test := pm.pipelines[hashstring]
+	if test {
+		return true
 	}
+	return false
+}
 
-	processing, err := processing.MakeDefaultSupport(false)(info, log, cfg)
-	if err != nil {
-		return nil, errors.Wrap(err, "error in MakeDefaultSupport")
-	}
+// registerPipeline is a small atomic function that registers a new pipeline with the managers
+// TODO: What happens if we try to register a pipeline that already exists? Which pipeline "wins"?
+func (pm *PipelineManager) registerPipeline(pipeline *Pipeline, hashstring string) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.pipelines[hashstring] = pipeline
 
-	pipelineCfg := pipeline.Config{}
-	err = cfg.Unpack(&pipelineCfg)
-	if err != nil {
-		return nil, errors.Wrap(err, "error unpacking pipeline config")
-	}
+}
 
-	idx, err := idxmgmt.DefaultSupport(log, info, config.Output.Config())
-	if err != nil {
-		return nil, errors.Wrap(err, "error making index manager")
-	}
+// removePipeline removes a pipeline from the manager if it's refcount is zero.
+func (pm *PipelineManager) removePipelineIfNeeded(hash string) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
 
-	pipeline, err := pipeline.Load(info,
-		pipeline.Monitors{
-			Metrics:   nil,
-			Telemetry: nil,
-			Logger:    log,
-		},
-		pipelineCfg,
-		processing,
-		func(stat outputs.Observer) (string, outputs.Group, error) {
-			cfg := config.Output
-			out, err := outputs.Load(idx, info, stat, cfg.Name(), cfg.Config())
-			return cfg.Name(), out, err
-		},
-	)
-	if err != nil {
-		return nil, errors.Wrap(err, "error in pipeline.Load")
+	//if the pipeline is no longer in use, clean up
+	if pm.pipelines[hash].refCount < 1 {
+		pipeline := pm.pipelines[hash].pipeline
+		delete(pm.pipelines, hash)
+		//pipelines must be closed after clients
+		//Just do this here, since the caller doesn't know if we need to close the libbeat pipeline
+		go func() {
+			pm.Logger.Debugf("Pipeline closing from removePipelineIfNeeded")
+			pipeline.Close()
+		}()
 	}
-
-	return pipeline, nil
 }
 
-// parseCfgKeys helpfully parses the values in the map, so users can specify yml structures.
-func parseCfgKeys(cfg map[string]string) (map[string]interface{}, error) {
+// registerClient registers a new client with the manager. Up to the caller to  actually close the libbeat client
+func (pm *PipelineManager) registerClient(cl *ClientLogger, hashstring, clientFile string) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	pm.clients[clientFile] = cl
+	pm.pipelines[hashstring].refCount++
+}
 
-	outMap := make(map[string]interface{})
+// removeClient deregisters a client
+func (pm *PipelineManager) removeClient(file string) (*ClientLogger, error) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
```

Comment:
I think will unlock an unlocked mutex from line 175 with a `fatal error: sync: unlock of unlocked mutex`

Reply:
Good catch!